# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Функция категоризации по детям
def children_category(children):
    if 1 <= children <= 2:
        return '1_2'
    if children >= 3:
        return 'многодетный'
    return 'нет детей' # Определим категории по детям
#Добавим столбец с категориями
data['children_rank'] = data['children'].apply(children_category) 
#Выберем нужные нам для анализа столбцы в новую таблицу
debt_children = data.pivot_table(index='children_rank', columns='debt', values='gender', aggfunc='count') 
debt_children.columns = ['no_debt', 'debt']
#Посчитаем долю должников
debt_children['reselt'] = debt_children['debt'] / (debt_children['debt'] + debt_children['no_debt']) 
print(debt_children)

               no_debt  debt    reselt
children_rank                         
1_2               6222   638  0.093003
многодетный        349    31  0.081579
нет детей        13028  1063  0.075438


**Вывод:** Как мы можем видеть, наличие детей влияет на уровень просроченной задолженности. 
У заемщиков без детей доля задолженности 0.075438, в то время, как у заемщиков с детьми 0.174582. Беспорно, в бюджете семьи с детьми имеются дополнительные статьи затрат "питание, памперсы, игрушки, развивашки". 
При этом размер задолженности у многодетных заемщиков ниже, чем в семьях с 1-2 детьми. Могу предположить, что многодетность дисциплинирует бОлее родителей, но и государство предоставляет больше льгот именно семьям от 3-х детей.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
#Выберем нужные нам столбцы
debt_status = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count') 
debt_status.columns = ['no_debt', 'debt']
# Посчитатем долю должников
debt_status['reselt'] = debt_status['debt'] / (debt_status['debt'] + debt_status['no_debt']) # Посчитатем долю должников
print(debt_status)

                       no_debt  debt    reselt
family_status                                 
Не женат / не замужем     2523   273  0.097639
в разводе                 1105    84  0.070648
вдовец / вдова             888    63  0.066246
гражданский брак          3749   385  0.093130
женат / замужем          11334   927  0.075606


**Вывод:** 
Согласно полученному результату, не женатые и состоящие в гражданском браке допускают "просрочку" чаще, чем остальные категории. 
Возможно в самом названии категории и таится ответ на вопрос "почему". Отстутствие желания брать ответственность за партнера или вообще его отсутствие говорит заранее о безответственном отношении человека ко всему, а как заемщика и к задолженности. #😜, а что))) кто из психологов изучил такой вопрос, может так и есть))))
Наименьшая доля простроченной задолженности допускают заемщики из категории "вдовец/вдова" #тут то какие ж могут быть шутки
Заемщики из категорий "женат/замужем" и "в разводе" показали почти одинаковый уровень стать должниками


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
debt_total_income = data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')
debt_total_income.columns = ['no_debt', 'debt']
debt_total_income['reselt'] = debt_total_income['debt'] / (debt_total_income['debt'] + debt_total_income['no_debt'])
print(debt_total_income) # Ваш код будет здесь. Вы можете создавать новые ячейки.

                       no_debt  debt    reselt
total_income_category                         
A                           23     2  0.080000
B                         4660   354  0.070602
C                        14568  1353  0.084982
D                          328    21  0.060172
E                           20     2  0.090909


**Вывод:** 
Ранее мы делали ранжирование по столбцу total_income_category, которым и воспользовались в этом исследовании.
Видим, что наибольшая доля задолженности проходиться на категорию Е с уровнем дохода 0-30000, что заранее и можно было предположить при рассмотрении заявки на кредит. Но доля в кредитном портфеле таких заемщиков невысока и делать обобщающие выводы было бы не совсем верно.
А вот следующие, с долей 0.084982, "С", с уровнем дохода от 50001 до 200000 могут дать нам ясную картину в исследновании. 
Именно с таким уровнем дохода больше всего и выдают заемные средства. Что и показывает уровень просроченной задолженности.


Я бы оветила так, что у нас есть понимание размера допущения неуплаты кредита в рамках нашего "от-до" уровня дохода, но точно не их зависимость. Как можно сделать выводы по вопросу  "зависимость между уровнем дохода и возвратом кредита в срок", если остальные данные составляют 25% - это 200000-1000000 (зачем они вообще брали кредит), а остальные - незначительный %% в исследовании. !!! Может это не "ревьюер душнила".....
В итоге ответ такой: мы опеределили зависимость между уровнем дохода и возвратом кредита в срок. Зависимость есть.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
debt_purpose_category = data.pivot_table(index='purpose_category', columns='debt', values='gender', aggfunc='count')
debt_purpose_category.columns = ['no_debt', 'debt']
debt_purpose_category['reselt'] = debt_purpose_category['debt'] / (debt_purpose_category['debt'] + debt_purpose_category['no_debt'])
print(debt_purpose_category)# Ваш код будет здесь. Вы можете создавать новые ячейки.

                          no_debt  debt    reselt
purpose_category                                 
операции с автомобилем       3879   400  0.093480
операции с недвижимостью     9971   780  0.072551
получение образования        3619   369  0.092528
проведение свадьбы           2130   183  0.079118


**Вывод:** Категории "операции с автомобилем" и "получение образования" показали самые высокие доли в просроченной задолженности. Можно предположить в силу того, что "операции с автомобилем" оформляют "бездумно" с силу обеспеченности кредита, автомобили же сразу принимаются в залог. А категория "получение образования" показывает высокую долю в связи с необеспеченностью заемщиков, они же только получают образование и работу заведомо не имеют.


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски информации были допущены в столбцах "days_employed" и в "total_income". Если опустить "человеческий фактор", то можно допустить, что это связано с тем, что многие займы выдают без запроса о трудовом стаже или типе занятости. Основной фактор, влияющий на выдачу кредита, отражен в столбце "total_income". 
Если бы доля пустых строк с данными о стаже или занятости была существенной для решения нашего иследования, то было бы приемлеммым проставить самим значения, опираясь на возраст заемщика, а не удалять строки.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Я пока такого преимущества не понимаю. Среднее арифметическое мне кажется вернее, с точки зрения математики.
Нам не хватаете данных, мы имеем выборку и проставляем в нее расчетное значение. Это мне кажется "ближе" к данным, для дальнейших арифметических действий или ранжирования. 
А медиальное значение, кажется мне, "что нам стоит дом построить, нарисуем, будем жить", просто число, делящее выборку пополам. Так давайте не пополам, а на треть, или четверть...или какое число кто любит. Не кажется мне это "полезным" для исследований данных, в которые мы проставим такие цифры.  

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

1.Наличие детей влияет на уровень просроченной задолженности. 
У заемщиков без детей доля задолженности 0.075438, в то время, как у заемщиков с детьми 0.174582. Беспорно, в бюджете семьи с детьми имеются дополнительные статьи затрат "питание, памперсы, игрушки, развивашки". 
При этом размер задолженности у многодетных заемщиков ниже, чем в семьях с 1-2 детьми. Могу предположить, что многодетность дисциплинирует бОлее родителей, но и государство предоставляет больше льгот именно семьям от 3-х детей.

2.Не женатые и состоящие в гражданском браке допускают "просрочку" чаще, чем остальные категории. 
Возможно в самом названии категории и таится ответ на вопрос "почему". Отстутствие желания брать ответственность за партнера или вообще его отсутствие говорит заранее о безответственном отношении человека ко всему, а как заемщика и к задолженности. 
Наименьшая доля простроченной задолженности допускают заемщики из категории "вдовец/вдова" #тут то какие ж могут быть шутки
Заемщики из категорий "женат/замужем" и "в разводе" показали почти одинаковый уровень стать должниками.

3.Наибольшая доля задолженности проходиться на категорию Е с уровнем дохода 0-30000, что заранее и можно было предположить при рассмотрении заявки на кредит. Но доля в кредитном портфеле таких заемщиков невысока и делать обобщающие выводы было бы не совсем верно.
А вот следующие, с долей 0.084982, "С", с уровнем дохода от 50001 до 200000 могут дать нам ясную картину в исследновании. 
Именно с таким уровнем дохода больше всего и выдают заемные средства. Что и показывает уровень просроченной задолженности.

4.Категории "операции с автомобилем" и "получение образования" показали самые высокие доли в просроченной задолженности. Можно предположить в силу того, что "операции с автомобилем" оформляют "бездумно" с силу обеспеченности кредита, автомобили же сразу принимаются в залог. А категория "получение образования" показывает высокую долю в связи с необеспеченностью заемщиков, они же только получают образование и работу заведомо не имеют.

5.Пропуски информации были допущены в столбцах "days_employed" и в "total_income". Если опустить "человеческий фактор", то можно допустить, что это связано с тем, что многие займы выдают без запроса о трудовом стаже или типе занятости. Основной фактор, влияющий на выдачу кредита, отражен в столбце "total_income".

6.Пока я не понимаю такого "лучшего решения для количественных переменных".

